### PR TITLE
[ZEPPELIN-1074][ZEPPELIN-1811] Enhance SSL Support

### PR DESCRIPTION
**Pull request by lresende**
### What is this PR for?

- Run SLL on the defined SSL port
- If Client Authentication is not enabled does not use it's properties
- Update Install documentation with instructions to enable SSL

### What type of PR is it?
[Bug Fix]

### What is the Jira issue?
* [ZEPPELIN-1074](https://issues.apache.org/jira/browse/ZEPPELIN-1074)
* [ZEPPELIN-1811](https://issues.apache.org/jira/browse/ZEPPELIN-1811)

### CODE REVIEW

**Comment by vinayshukla** [comment]
@lresende In case the password is obfuscated by Jetty's tool, is there any code required to de-obfuscate the password before passing it onto SSLContext?

File: conf/README.md [hunk 1]
```diff
@@ -13,3 +15,86 @@ java -cp ./zeppelin-server/target/lib/jetty-all-server-<version>.jar org.eclipse
 
 If you are using a self-signed, a certificate signed by an untrusted CA, or if client authentication is enabled, then the client must have a browser create exceptions for both the normal HTTPS port and WebSocket port. This can by done by trying to establish an HTTPS connection to both ports in a browser (i.e. if the ports are 443 and 8443, then visit https://127.0.0.1:443 and https://127.0.0.1:8443). This step can be skipped if the server certificate is signed by a trusted CA and client auth is disabled.
 
+### Configuring server side SSL
+
+The following properties needs to be updated in the **zeppeling-env.xml** in order to enable server side SSL.
```
Comment (FireArrow) [inline conf/README.md]: The file in question is `zeppelin-site.xml`

File: conf/README.md [hunk 2]
```diff
@@ -13,3 +15,86 @@ java -cp ./zeppelin-server/target/lib/jetty-all-server-<version>.jar org.eclipse
...
+  <name>zeppelin.ssl.key.manager.password</name>
+  <value>change me</value>
+  <description>Key Manager password. Defaults to keystore password. Can be obfuscated.</description>
+</property>
+```
+
+
+### Enabling client side certificate authentication
+
+The following properties needs to be updated in the **zeppeling-env.xml** in order to enable client side certificate authentication.
```
Comment (FireArrow) [inline conf/README.md]: The file in question is `zeppelin-site.xml`

File: conf/README.md [hunk 3]
```diff
@@ -13,3 +15,86 @@ java -cp ./zeppelin-server/target/lib/jetty-all-server-<version>.jar org.eclipse
...
+
+```
+<property>
+  <name>zeppelin.server.ssl.port</name>
+  <value>8443</value>
+  <description>Server ssl port. (used when ssl property is set to true)</description>
+</property>
```
Comment (FireArrow) [inline conf/README.md]: This property is specified both here and in the section above, even though no other property is duplicated between sections

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java [hunk 4]
```diff
@@ -314,6 +318,9 @@ public String getKeyManagerPassword() {
 
   public String getTrustStorePath() {
     String path = getString(ConfVars.ZEPPELIN_SSL_TRUSTSTORE_PATH);
+    if (path == null) {
+      path = getKeyStorePassword();
```
Comment (FireArrow) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java]: Shouldn't this be `getKeyStorePath()`? And that should mean that if keystore is used as truststore then `getTrustStorePassword()` need the same logic

Comment (lresende) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java]: Rhanks @FireArrow fixed getTrustStorePath, and getTrustStorePassword was already defaulting to getKeyStorePassword

**Comment by lresende** [comment]
@vinayshukla I have extended the README.md to detail how to obfuscate the passwords using the Jetty password tool

**Comment by lresende** [comment]
@Leemoonsoo Could you please take a quick look at this. Thank you.

**Comment by Leemoonsoo** [comment]
Thanks @lresende for enhancement!
Looks good to me.

**Comment by felixcheung** [comment]
LGTM!

**Comment by felixcheung** [comment]
merging if no more comment

**Comment by AhyoungRyu** [comment]
Sorry for my late blocking, but actually I even didn't know that there is `README.md` under `conf/`. Maybe many ppl are same with me I guess. Seems there is already some duplicated information existed. How about adding this file's location using hyperlink (or whatever) under [this section](https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/install/install.html#apache-zeppelin-configuration)? It would be better to be provided in official docs since this is useful info.

**Comment by lresende** [comment]
No worries @AhyoungRyu I will make the suggested changes and add ZEPPELIN-1811 to this pr as well.

**Comment by lresende** [comment]
@AhyoungRyu @felixcheung Please take a look and let me know if I should update any other pieces.

**Comment by AhyoungRyu** [comment]
@lresende Thanks! Looks nice. BTW the [Build from Source](https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/install/install.html#building-from-source) section went to too down side. How about putting it in the above [Apache Zeppelin configuration](https://zeppelin.apache.org/docs/0.7.0-SNAPSHOT/install/install.html#apache-zeppelin-configuration) section instead? 

Other than that LGTM 👍

**Comment by felixcheung** [comment]
looks like we only have a minor comment - could we update and then we could merge?

**Comment by AhyoungRyu** [comment]
@felixcheung Yes. So I can handle it later if you don't mind. I think it'll be faster. Shall we merge then? :)

**Comment by lresende** [comment]
@felixcheung and @AhyoungRyu Sorry for the delay, as I was a little offline in the last few days. I have implemented the suggestion and this should be ready to merge.

**Comment by AhyoungRyu** [comment]
@lresende Thanks! Will check again.

**Comment by AhyoungRyu** [comment]
@lresende Looks good 👍 
Merge this into master if there are no more comments.

**Comment by AhyoungRyu** [comment]
CI failure is not related with this change and already reported in [ZEPPELIN-1857](https://issues.apache.org/jira/browse/ZEPPELIN-1857). So I'm merging this now :)

**Pull request closed without merging**